### PR TITLE
Set shuffle to DataPipes with set_shuffle API

### PR DESCRIPTION
**Pull request by ejguan**
This PR requires PR is landed: https://github.com/pytorch/pytorch/pull/83202

## changes
- For `apply_shuffle_setting` and `apply_shuffle_seed`, it makes sure it will apply shuffle setting to each of DataPipe that contains a method called `set_shuffle` or `set_seed`.
- Change the API from `apply_shuffle_seed` to `apply_random_seed`.
- Fix a bug that `apply_shuffle_seed` only accepts DataPipe that is hashable. After the PR, this function uses `id` to prevent seeding the same DataPipe multiple times per epoch.
- Fix another bug from `shuffler` that `reset` with `_enable=False` would also reset `_seed`.

### CODE REVIEW

File: torch/utils/data/graph_settings.py [hunk 1]
```diff
@@ -66,13 +83,46 @@ def apply_shuffle_settings(datapipe: DataPipe, shuffle: Optional[bool]) -> DataP
...
+
 def apply_shuffle_seed(datapipe: DataPipe, rng: Any) -> DataPipe:
+    warnings.warn(
+        "`apply_shuffle_seed` is deprecated since 1.13 and will be removed in the future releases.\n"
+        "Plese use `apply_random_seed` to set random seeds for the graph of `DataPipes`."
+    )
+    return apply_random_seed(datapipe, rng)
+
+
+def apply_random_seed(datapipe: DataPipe, rng: torch.Generator) -> DataPipe:
```
Comment (ejguan) [inline torch/utils/data/graph_settings.py:102]: Anyone has a better name? like `set_random_seed` or `reset_random_seed`?

File: torch/utils/data/graph_settings.py [hunk 2]
```diff
@@ -45,13 +47,29 @@ def apply_sharding(datapipe: DataPipe, num_of_instances: int, instance_id: int)
...
+        return False
+    if not inspect.ismethod(datapipe.set_shuffle) or not inspect.ismethod(datapipe.set_seed):
+        return False
+    return True
+
+
+def apply_shuffle_settings(datapipe: DataPipe, shuffle: Optional[bool] = None) -> DataPipe:
+    r"""
+    Traverse the graph of ``DataPipes`` to find and set shuffle attribute
+    to each `DataPipe` that has an API of ``set_shuffle``.
```
Comment (NivekT) [inline torch/utils/data/graph_settings.py:61]: If I'm reading `_is_shuffle_datapipe` correctly, it requires both ``set_shuffle`` and ``set_seed`` to exist right?
```suggestion
    to each `DataPipe` that has an API of ``set_shuffle`` and ``set_seed``.
```

**Review by NivekT — APPROVED** [review]
LGTM! I think we should find a place (probably in the doc) to communicate with users about these randomness methods (if we expect users to interact with them) and what the requirements are for custom DataPipes (e.g. having a method named `set_seed`) and etc.

**Comment by ejguan** [comment]
> I think we should find a place (probably in the doc) to communicate with users about these randomness methods (if we expect users to interact with them) and what the requirements are for custom DataPipes (e.g. having a method named `set_seed`) and etc.

I will probably do it when we have `DataLoader2` with the random control enabled.
